Added server-settings.json usage

hub23 opened the pull request:
As stated in [Including server-settings.json #19](https://github.com/zopanix/docker_factorio_server/issues/19), I would like to use the settings defined in `server-settings.json`. The needed settings can not be called by a parameter passed to the server binary, the file `server-settings.json` must be used to set the values.

Including these changes, `/opt/factorio/server-settings.json` will be generated and populated by `cat` in `new_smart_launch.sh`. If one or more variables are set while starting the container, the `server-settings.json` config file will be used to define additional settings by adding `--server-settings /opt/factorio/server-settings.json` to `$factorio_command`.

The following parameters are passable at container start to define settings:

```
FACTORIO_SERVER_NAME
FACTORIO_SERVER_DESCRIPTION
FACTORIO_SERVER_MAX_PLAYERS
FACTORIO_SERVER_VISIBILITY
FACTORIO_USER_USERNAME (needed if visibility is set to public)
FACTORIO_USER_PASSWORD (neede if visibility is set to public)
FACTORIO_USER_TOKEN (can be used instead of password)
FACTORIO_SERVER_GAME_PASSWORD (define server password)
FACTORIO_SERVER_VERIFY_IDENTITY (requires factorio.com account for user at login)
```

Conditions for execution of:
`factorio_command="$factorio_command --server-settings /opt/factorio/server-settings.json"`

Are the following:

```
if [ "$FACTORIO_SERVER_NAME" ] \
|| [ "$FACTORIO_SERVER_DESCRIPTION" ] \
|| [ "$FACTORIO_SERVER_MAX_PLAYERS" ] \
|| [ "$FACTORIO_SERVER_VISIBILITY" ] \
|| [ "$FACTORIO_USER_USERNAME" ] \
|| [ "$FACTORIO_USER_PASSWORD" ] \
|| [ "$FACTORIO_USER_TOKEN" ] \
|| [ "$FACTORIO_SERVER_GAME_PASSWORD" ] \
|| [ "$FACTORIO_SERVER_VERIFY_IDENTITY" ]
```

The defaults (`/opt/factorio/data/server-settings.example.json` in `zopanix/factorio:0.13.8`) for the `server-settings.json` to contain are:

```
  "name": "Name of the game as it will appear in the game listing",
  "description": "Description of the game that will appear in the listing",
  "tags": ["game", "tags"],
  "max_players": "0",
  "visibility": "public",
  "username": "",
  "password": "",
  "game_password": "",
  "verify_user_identity": true
```

Some lines are hidden in this comment because of comments and braces.

If default values are missing from `server-settings.json`, the server wont start (Although I did not test one by one, just tested by removing "name", I assume all defaults are needed.).

I changed the following values away from default because I think default factorio.com account verification for user login or public listing, which also requires factorio.com account, is not wanted by most users:

```
  "name": "Factorio Server $VERSION"
  "visibility": "hidden",
  "verify_user_identity": false
```

Required credentials for public listing can be supplied by setting `FACTORIO_USER_USERNAME` and `FACTORIO_USER_PASSWORD` or `FACTORIO_USER_TOKEN` variables. A check for public visibility without `FACTORIO_USER_USERNAME` and `FACTORIO_USER_PASSWORD` or `FACTORIO_USER_TOKEN` defaults back to "hidden" when parameters are missing.

hub23 committed:
Added server-settings.json usage

server-settings.json will be generated and populated by cat in new_smart_launch.sh.
The following parameters are passable at container start to define settings:

FACTORIO_SERVER_NAME
FACTORIO_SERVER_DESCRIPTION
FACTORIO_SERVER_MAX_PLAYERS
FACTORIO_SERVER_VISIBILITY
FACTORIO_USER_USERNAME (needed if visibility is set to public)
FACTORIO_USER_PASSWORD (neede if visibility is set to public)
FACTORIO_USER_TOKEN (can be used instead of password)
FACTORIO_SERVER_GAME_PASSWORD (define server password)
FACTORIO_SERVER_VERIFY_IDENTITY (check for factorio.com account)

If one or more variables are set while starting the container, the server-settings.json will be used.

Conditions for execution of:
factorio_command="$factorio_command --server-settings /opt/factorio/server-settings.json"

Are the following:
if [ "$FACTORIO_SERVER_NAME" != factorio server ] || [ -n "$FACTORIO_SERVER_DESCRIPTION" ] || [ -n "$FACTORIO_SERVER_MAX_PLAYERS" ] || [ "$FACTORIO_SERVER_VISIBILITY" != hidden ] || [ -n "$FACTORIO_USER_USERNAME" ] || [ -n "$FACTORIO_USER_PASSWORD" ] || [ -n "$FACTORIO_USER_TOKEN" ] || [ -n "$FACTORIO_SERVER_GAME_PASSWORD" ] || [ "$FACTORIO_SERVER_VERIFY_IDENTITY" != false ]

The defaults (`/opt/factorio/data/server-settings.example.json` in `zopanix/factorio:0.13.8`) for the `server-settings.json` to contain are:
```
  "name": "Name of the game as it will appear in the game listing",
  "description": "Description of the game that will appear in the listing",
  "tags": ["game", "tags"],
  "max_players": "0",
  "visibility": "public",
  "username": "",
  "password": "",
  "game_password": "",
  "verify_user_identity": true
```
Some comment values and braces are hidden in this comment.

If default values are missing from `server-settings.json`, the server wont start (Although I did not test one by one, just tested by removing "name", I assume all defaults are needed.).

I changed the following values away from default because I think default factorio.com account verification or public listing is not wanted by most users.
```
  "visibility": "hidden",
  "verify_user_identity": false
```

Needed credentials for public listing can be supplied by setting `FACTORIO_USER_USERNAME` and `FACTORIO_USER_PASSWORD` or `FACTORIO_USER_TOKEN` variables.

hub23 committed:
Fixed Condition

Fixed if condition for execution of --server-settings factorio command

hub23 committed:
server-settings.jsoc echo output

Added server-settings.json output via echo and some fixes.

hub23 committed:
Moved server-settings.json output below rcon password output.

hub23 committed:
Merged ENV entrys.

hub23 committed:
Added README entrys

hub23 committed:
Restructured "server-settings.json inclusion if condition" at line 68 to use a new line for every statement.

hub23 committed:
Shrinked the README size by combining factorio.com auth login, password and token example.

hub23 committed:
Added additional descriptions to README.md

psykzz commented:
You should remove the lineending and filespace changes.
No need to modify them just to suit style as it can disturbs `git blame`.

psykzz commented on new_smart_launch.sh:
```diff
@@ -10,15 +10,41 @@ echo '  \  \::/     \  \::/       \  \:\  /:/  \__\/  \:\   \  \:\  /:/   \  \::
+  "_comment_visibility": ["public: Game will be published on the official Factorio matching server",
+                          "lan: Game will be broadcast on LAN",
+                          "hidden: Game will not be published anywhere"],
+  "visibility": "$FACTORIO_SERVER_VISIBILITY",
```
Might want to do a preemptive check for public without a username / password and default back to "hidden", maybe also alert the user that this is the reason its happening.

hub23 replied:
Check added.

psykzz commented on Dockerfile:
```diff
@@ -17,7 +17,16 @@ ENV FACTORIO_AUTOSAVE_INTERVAL=2 \
+    FACTORIO_USER_TOKEN="" \
+    FACTORIO_SERVER_GAME_PASSWORD="" \
+    FACTORIO_SERVER_VERIFY_IDENTITY="false"
 
```
Instead of checking defaults and doing a check later on to determine if "not default" why not set empty here and set default later in the smart_launch.sh.

hub23 replied:
I changed the check logic.

psykzz commented on new_smart_launch.sh:
```diff
@@ -55,6 +94,19 @@ then
 factorio_command="$factorio_command --rcon-password $FACTORIO_RCON_PASSWORD"
+# Show server-settings.json config
+echo "###"
+echo "# server-settings.json config:"
```
Could just call this `Server Config` instead of being so explicit.

hub23 replied:
Changed the description and sorting.

psykzz commented on new_smart_launch.sh:
```diff
@@ -34,10 +60,23 @@ fi
 fi
+# Include server-settings.json if one or more variables are populated
+if [ "$FACTORIO_SERVER_NAME" != "factorio server" ] \
+|| [ -n "$FACTORIO_SERVER_DESCRIPTION" ] \
```
`-n` isn't exactly required in these tests according to http://stackoverflow.com/questions/3869072/test-for-non-zero-length-string-in-bash-n-var-or-var

hub23 replied:
Removed unnecessary parameters.

psykzz commented:
Other then some small comments, 👍 Saves having to hack together a solution by polluting an environment variable.

hub23 committed:
removing condition parameters

removing -n parameters from server-settings.json if condition because they are not required

hub23 committed:
undo style changes

hub23 commented:
> You should remove the lineending and filespace changes.

I'm pretty new to git, that makes sense :grin:.
I reverted the changes in new_smart_launch.sh.

hub23 committed:
Changed Info Description at server-settings.json config echo

hub23 committed:
Changed server-settings var check logic

Changed the check for user var logic for server-settings.json as requested here:
https://github.com/zopanix/docker_factorio_server/pull/21/files#r71088761

hub23 committed:
checking for credentials

checking for credentials if visibility is set to public, default to hidden if not supplied
restructured server-settings part as child of conditions

hub23 commented:
I will update here if new changes are tested.

hub23 committed:
condition fix

hub23 committed:
condition fix

hub23 committed:
fixes

hub23 committed:
fix

hub23 committed:
fixes

hub23 committed:
Added echo statement

hub23 committed:
fix

hub23 committed:
fix

hub23 committed:
fixed echo spelling

hub23 committed:
added auth error info to README.md

hub23 committed:
minor changes

hub23 committed:
change default for Server Name from "Factorio Server" to "Factorio Server $VERSION"

psykzz commented:
LGTM :+1:

hub23 committed:
commented out token parts because of:
https://github.com/zopanix/docker_factorio_server/issues/23

hub23 committed:
readme comment fix

hub23 committed:
comment syntax fix

hub23 committed:
added README comment

hub23 commented:
Tested a few cases and almost everything is working :+1: [except](https://github.com/zopanix/docker_factorio_server/issues/23) factorio.com login token.

hub23 committed:
Commented out token related line

hub23 committed:
comment fix

zopanix commented:
hey, I'm accepting this bud only had time to read it quickly
